---
title: "evaluate mock communities"
author: "Kimberly Ledger"
date: "2024-11-08"
output: html_document
---

load libraries 
```{r, message=FALSE}
library(tidyverse)
library(dplyr)
rename <- dplyr::rename
```

first, i'm going to compare the mock community asvs from the filtered.seqTab output and the decontaminated output (just out of curiosity)


load ASV table and metadata - this the output from dadasnake
```{r}
asv_table <- readRDS("/home/kimberly.ledger/nearshore_eDNA/data/samples_w_mocks_updated/filtered.seqTab.RDS") %>%
  select(!Row.names)

#transpose 
asv_table <- data.frame(t(asv_table))

#set column names to be ASV# 
colnames(asv_table) <- asv_table["ASV",]

#remove row that has ASV#
asv_table <- asv_table[!rownames(asv_table) %in% c('ASV'), ]

#make sure reads are numbers
# Convert all character columns to numeric
for (col in names(asv_table)) {
  asv_table[[col]] <- as.numeric(asv_table[[col]])
}

#make make sample ID a column 
asv_table$Sample_ID <- rownames(asv_table)

asv_table <- asv_table %>%
  select(Sample_ID, everything()) %>%
  mutate(Sample_ID = str_replace(Sample_ID, "^COM-", "COM"))
```

filter asv table to just include the mock community samples 
```{r}
asv_first <- which(startsWith(names(asv_table), "ASV"))[1]
asv_last <- ncol(asv_table)

mock_table_original <- asv_table %>%
  filter(grepl("COM", Sample_ID)) %>%
  pivot_longer(cols = c(asv_first:asv_last), names_to = "ASV", values_to = "reads") %>%
  filter(reads > 0)
```

load decontaminated reads 
```{r}
decontaminated_table <- read.csv("/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mocks_updated/decontaminated_asv_table.csv") %>%
  select(!X) %>%
  filter(grepl("COM", Sample_ID)) %>%
  select(!sample_type) %>%
  filter(reads > 0)
```

anti-join tables 
```{r}
comparison <- mock_table_original %>%
  full_join(decontaminated_table, by = c("Sample_ID", "ASV"))
```

alright, going to use the decontaminated read output!

load mock community metadata
```{r}
#read in some sample metadata to work with 
mock_metadata <- read.csv("/home/kimberly.ledger/nearshore_eDNA/data/samples_w_mocks_updated/nearshore_mock_compositions_w_conc.csv") %>%
  mutate(replicate = "expected") %>%
  mutate(sample_ID = paste0(Community_ID, "-E")) %>%
  group_by(Community_ID) %>%
  mutate(start_conc_ng = uL * conc_ng_uL_measured,
         community_total_conc = sum(start_conc_ng),
         Perc_Community = start_conc_ng/community_total_conc)

#rename some taxa so that the mock community expected match the taxonomic assignments
mock_metadata <- mock_metadata %>%
  mutate(taxon = ifelse(species == "Sebastes flavidus", "Sebastes 3", taxon),
         taxon = ifelse(species == "Sebastes auriculatus", "Sebastes 2", taxon),
         taxon = ifelse(species == "Sebastes caurinus", "Sebastes 1", taxon),
         taxon = ifelse(species == "Limanda aspera", "Pleuronectidae 1", taxon))
```

assign asv's to taxons 
```{r}
taxonomy <- read.csv("/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mocks_updated/taxonomy_20241112_collapsed.csv") %>%
  select(!X) %>%
  select(qseqid, taxon) %>%
  unique() %>%
  rename(ASV = qseqid) 

mock_taxons <- decontaminated_table %>%
  left_join(taxonomy, by = "ASV")
```

take a closer look at what asv's make up the taxons
```{r}
mock_taxons %>%
  group_by(taxon) %>%
  mutate(taxon_reads = sum(reads)) %>%
  group_by(ASV, taxon, taxon_reads) %>%
  summarize(tot_reads = sum(reads)) %>%
  mutate(prop_asv = tot_reads/taxon_reads) %>%
  ggplot(aes(x=taxon, y=prop_asv, fill=ASV)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~Community_ID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "taxon",
    title = "") + 
  theme(
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

make list of just the ASVs found in mock community samples 
```{r}
asvs_in_mocks <- mock_taxons %>%
  group_by(ASV, taxon) %>%
  summarise(tot_reads = sum(reads))
```

## i also ran blastn on the filtered.seqs.fasta against a reference db with just the mock community species 
```{r}
mock_assignments <- read.csv("/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mocks_updated/mockdb_taxonomy.csv") %>%
  select(!X) %>%
  rename(ASV = qseqid)
```

compare 
```{r}
asvs_in_mocks <- asvs_in_mocks %>%
  left_join(mock_assignments)
```

```{r}
asvs_in_mocks <- asvs_in_mocks %>%
  filter(ASV != "ASV_0610") %>%         ## remove the asvs that don't belong 
  filter(ASV != "ASV_0864") %>% 
  filter(ASV != "ASV_0853") %>%
  select(ASV, taxon, tot_reads)

write.csv(asvs_in_mocks, "/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mocks_updated/asvs_in_mocks.csv")
```



very similar, going to continue with the blastn taxonomic assignments

reformat and join the mock reads with expected mock values 
```{r}
mock_table <- mock_taxons %>%
  filter(ASV != "ASV_0610") %>%         ## remove the asvs that don't belong 
  filter(ASV != "ASV_0864") %>% 
  filter(ASV != "ASV_0853") %>%
  separate(Sample_ID, into = c("Community_ID", "replicate"), sep = "-", remove = FALSE) %>%
  group_by(Sample_ID, Community_ID, replicate, taxon) %>%
  summarize(reads = sum(reads)) %>%
  ungroup() %>%
  group_by(Sample_ID, Community_ID, replicate) %>%
  mutate(community_reads = sum(reads)) %>%
  mutate(Perc_Community = reads/community_reads)

mock_all <- mock_table %>%
  bind_rows(mock_metadata)
```


export the mock community data in the format needed for quantitative metabarcoding models
```{r}
mock_start <- mock_metadata %>%
  select(Community_ID, taxon, start_conc_ng) %>%
  rename(community = Community_ID)

mock_export <- mock_table %>%
  select(Community_ID, replicate, taxon, reads) %>%
  rename(community = Community_ID) %>%
  rename(tech_rep = replicate) %>%
  rename(nReads = reads) %>%
  mutate(Cycles = 35) %>%
  left_join(mock_start) %>%
  mutate(start_conc_ng = ifelse(is.na(start_conc_ng), 0, start_conc_ng))

write.csv(mock_export, "/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mocks_updated/ns_mock_20241112.csv", row.names = F)
```

plot the mock community reads counts
```{r}
mock_all %>%
  filter(!is.na(reads)) %>%
  ggplot(aes(x=Sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  facet_grid(~Community_ID, scales = 'free') +
  theme_bw() +
  labs(
    y = "sequencing reads",
    x = "sample",
    title = "") + 
  theme(
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

perhaps COM-5-B has too low of a read count to use? - establish 5000 read threshold???

now plot species proportions with expected 
```{r }
mock_all %>%
  filter(Community_ID == "COM1") %>%
  filter(Perc_Community > 0) %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM1 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM1") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```



now plot species proportions with expected 
```{r }
mock_all %>%
  filter(Community_ID == "COM2") %>%
  filter(Perc_Community > 0) %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM2 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM2") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM3") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM3 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```


```{r}
mock_all %>%
  filter(Community_ID == "COM3") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM4") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM4 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM4") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM5") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM5 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM5") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM6") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM6 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM6") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```

```{r }
mock_all %>%
  filter(Community_ID == "COM7") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM7 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```


```{r}
mock_all %>%
  filter(Community_ID == "COM7") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM8") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM8 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM8") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM9") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM9 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM9") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```



```{r }
mock_all %>%
  filter(Community_ID == "COM10") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM10 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM10") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM11") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM11 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM11") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```

```{r }
mock_all %>%
  filter(Community_ID == "COM12") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM12 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM12") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```


```{r }
mock_all %>%
  filter(Community_ID == "COM13") %>%
  ggplot(aes(x=Sample_ID, y=Perc_Community, fill=taxon)) +
  geom_bar(stat = "identity") + 
  #facet_grid(~SampleID, scales = 'free') +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample",
    title = "COM13 - proportion of reads") + 
  theme(
    #axis.text.x=element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.95),
    legend.text = element_text(size = 8),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    legend.title = element_blank()
  )
```

```{r}
mock_all %>%
  filter(Community_ID == "COM13") %>%
  select(Sample_ID, taxon, Perc_Community) %>%
  pivot_wider(names_from = taxon, values_from = Perc_Community)
```
