---
title: "Taxonomic Assignment"
author: "Kimberly Ledger"
date: "2024-07-17"
output: html_document
---

updated: 21 Aug 2024 

taxonomic assignment of nearshore eDNA ASVs from 20240701 sequencing run with mock communities samples (sequenced on 20240820) included 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load-libraries}
library(tidyverse)
library(tidyr)
library(stringr)
rename <- dplyr::rename
```

## load output from taxonkit 
```{r load-output-from-taxonkit}
taxonomy <-read.delim("/home/kimberly.ledger/nearshore_eDNA/data/samples_w_mock/blastn_taxlineage_ns_w_mock.txt", header = FALSE, na.strings=c(""," ","NA"))

head(taxonomy)
```


## clean up the header a bit
```{r}
# use the full taxonomy rather than the seq id to collapse identical entries
tax_df <- taxonomy %>%
  filter(V4 > 100) %>% # make sure all retained matches are >100 bp
  select(-V2, -V5, -V6, -V7, -V14, -V8, -V9, -V10, -V11, -V12, -V13, -V15, -V16) %>%  #remove unnecessary columns
  group_by(V1, V17) %>% # group by the sequence key and the full taxonomy to reduce duplicate entries
  unique() %>% # doing that reduced the number of entries
  rename(qseqid=V1, perc_id=V3, length=V4, taxonomy=V17) %>% #rename headers
  filter(perc_id > 96) %>% 
  filter(!is.na(taxonomy))
```

## calcualate the number of ASV's with taxonomic assignment
```{r how-many-total-ASVs-w-taxonomy}
tax_df %>% 
  ungroup() %>%
  select(qseqid) %>%
  unique() %>%
  nrow
```

## reformat table 
```{r}
# formatting the taxonomy variables
taxon_df <- tax_df %>%
  filter(str_detect(taxonomy, ";")) %>%
  separate(taxonomy, into=c("kingdom", "phylum", "class", "order", "family", "genus", "species"), sep = ";") %>% 
  filter(!grepl("sp\\.", species, ignore.case = TRUE)) %>%
  filter(!grepl("cf\\.", species, ignore.case = TRUE)) %>%
  filter(!grepl("aff\\.", species, ignore.case = TRUE)) %>%
  filter(!grepl(" x ", species, ignore.case = TRUE)) %>%         ### remove hybrids and unknown species assignments
  filter(!grepl("\\d", species))
```

## what non-fish are here? 
```{r}
not_Actinopteri <- taxon_df %>%
  filter(class != "Actinopteri") %>%
  select(species, genus, family, order, class, phylum) %>%
  unique()
```
for this project, non-fish are not of interest so we will just remove all of them from the analysis

## let's filter out species based on geographic distribution 

## now look at the fish and figure out what taxa are in/out of our range
```{r}
to_check_range <- taxon_df %>%
  filter(class == "Actinopteri") %>%
  select(species, genus, family, order, class, phylum) %>%
  unique()
```

## check ranges for species using rfishbase 
## also at this step, check 
```{r}
#library(remotes)
#remotes::install_github("ropensci/rfishbase")
library(rfishbase)

#one-time download of all fishbase tables... this takes a bit 
fb_import()

#first, validate species names using rfishbase synonyms
spp_df <- synonyms(to_check_range$species)

syn <- spp_df %>% 
  filter(Status == "synonym") %>%
  select(synonym, Species) %>%
  unique()
  
to_check_range_1 <- to_check_range %>%
  filter(!species %in% syn$synonym)

to_check_range <- c(to_check_range_1$species, syn$Species)

#get distribution info 
spp_distribution <- distribution(to_check_range) %>%
  select(Species, FAO) %>%
  unique()

#add column to designate if we will consider a species as "in range"- for this study, this will be NE Pacific
spp_distribution_range <- spp_distribution %>%
  mutate(in_range = ifelse(is.na(FAO), NA, "no"),
         in_range = ifelse(FAO == "Pacific, Northeast", "yes", in_range))

#keep just a list of spp names and yes/no/NA for "in range"  - this way we can keep track of what spp didn't have any reference information in fishbase to determine range 
range_na <- spp_distribution_range %>%
  select(Species, in_range) %>%
  unique() %>%
  filter(is.na(in_range))

range_na <- range_na %>%
  mutate(in_range = ifelse(Species == "Ammodytes japonicus", "no", in_range),
         in_range = ifelse(Species == "Sebastes cheni", "no", in_range))

in_range <- spp_distribution_range %>%
  select(Species, in_range) %>%
  filter(in_range == "yes")
```

When a valid name was not found, the presence of a species in the study area was checked using the GBIF database (https://www.gbif.org/).


# update taxon dataframe with correct species names (i.e. not synonyms) and then filter the dataframe
```{r}
## what species are in range but used a synonym in taxon dataframe? 
temp <- in_range %>%
  filter(Species %in% syn$Species)
temp 

#what names were used in the taxon_df? 
syn %>%
  filter(Species %in% temp$Species)

#rename these in the taxon_df
taxon_df_updated_syn <- taxon_df %>%
  mutate(species = ifelse(species == "Gadus ogac", "Gadus macrocephalus", species),
         species =  ifelse(species == "Polypera greeni", "Liparis greeni", species),
         genus = ifelse(genus == "Polypera", "Liparis", genus),
         species =  ifelse(species == "Embassichthys bathybius", "Microstomus bathybius", species),
         genus = ifelse(genus == "Embassichthys", "Microstomus", genus))

taxon_df_inrange <- taxon_df_updated_syn %>%
  filter(species %in% in_range$Species)
```


## i've identified a few additional species that are not in AK that removing can improve taxonomic assignment of some ASVs
```{r}
taxon_df_inrange <- taxon_df_inrange %>% 
  filter(species != "Sebastes rastrelliger") %>%    ## only of CA coast
  filter(species != "Hippoglossoides elassodon") %>%  ## might want to keep this 
  filter(species != "Hippoglossoides robustus") %>%
  mutate(species = ifelse(species == "Sebastes caurinus", "Sebastes caurinus or Sebastes maliger", species),
         species = ifelse(species == "Sebastes maliger", "Sebastes caurinus or Sebastes maliger", species))
```


## figure out what ASVs have only one taxon hit
```{r}
single_spp_seqs <- taxon_df_inrange %>% 
  group_by(qseqid) %>%
  add_tally(name = "n_taxa") %>%
  filter(n_taxa == 1) %>%
  rename(seq_percID = perc_id) %>%
  select(!length) %>%
  select(!n_taxa)
```

## remove the single-species seqs from the dataframe and then rank the hits by % identity for the remaining seqs
```{r}
seq_id_diff <- taxon_df_inrange %>%
  anti_join(., single_spp_seqs) %>%  ## comment out this line if there are no single species ASVs
  select(-length) %>%
  group_by(qseqid, species, genus, family, order, class, phylum, kingdom) %>%
    mutate(seq_percID = max(perc_id)) %>%
    group_by(qseqid, species, genus, family, order, class, phylum, kingdom, seq_percID) %>%
  summarise(max(seq_percID)) %>% # take just the top hit for each taxon (for each sequence)
  select(-`max(seq_percID)`) %>%
  ungroup() %>%
  group_by(qseqid) %>%
       #mutate(id_rank = rank(desc(seq_percID), ties.method = "min")) %>% # rank the taxonomic hits per sequence by % id
       mutate(top_perc = max(seq_percID)) %>% # designate the highest % id for the best taxonomic hit in each sequence (in some, but not all cases, this is 100%)   %>%
      mutate(diff = top_perc - seq_percID) %>% # calculate the difference between the % identity of the top hit and each subsequent taxonomic hit
        arrange(diff)
```

Now you have the single best entry for each species for each sequence ranked and with the difference between the first and second ranked entries calculated.

For sequences with multiple top hits, where the difference between ranked taxa = 0, you will end up defaulting to genus- or family-level ID. We'll do the same for any sequences where the difference between the first and second ranked taxa is < 1%.

## Figure out which differences are **> 0.001%** and eliminate those first - 
```{r}
# filter out any taxa that are >1% less matching identity than the top taxonomic hit for a given sequence
to_remove_low_perc_hits <- seq_id_diff %>%
  ungroup() %>%
  group_by(qseqid) %>%
  filter(diff > 0.001)

keepers <- seq_id_diff %>%
  anti_join(to_remove_low_perc_hits)
```

## so now determine whether ASVs should be assigned to genus, family, order, etc. 
```{r}
singletons <- keepers %>%
  select(qseqid) %>%
  tally() %>%
  filter(n == 1)

# these are the seqs that now have only a single match
singleton_df <- singletons %>%
  left_join(keepers) %>%
  select(-n) %>%
  bind_rows(single_spp_seqs) %>% # combine the single spp data
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)

## Genus-level matches
# remove the singletons from the bigger df 
single_genus <- keepers %>%
  anti_join(singleton_df)%>% # at best, these should be genus-level matches
  group_by(qseqid, genus) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single genus
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "genus") %>%
  mutate(taxon = genus)
  

## Family-level matches
single_family <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  group_by(qseqid, family) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single family
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "family") %>%
  mutate(taxon = family)


## Order-level matches
single_order <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  group_by(qseqid, order) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single order
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "order") %>%
  mutate(taxon = order)


## Class-level matches
single_class <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  group_by(qseqid, class) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single class
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "class") %>%
  mutate(taxon = class)


## Phylum-level matches
single_phylum <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, phylum) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single phylum
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "phylum") %>%
  mutate(taxon = phylum)

# Kingdom level matches
single_kingdom <- keepers %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, kingdom) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single kingdom
  select(-n) %>%
  left_join(., keepers) %>%
  mutate(taxonomic_level = "kingdom") %>%
  mutate(taxon = kingdom)
```

## modify the singleton_df to include the right variable headers
```{r}
single_spp <- singleton_df %>%
  #select(-perc_id, -length, -n_taxa) %>%
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)
```

## recombine the full data set now that the appropriate level of taxonomy has been determined
```{r}
sorted_tax_df <- bind_rows(single_kingdom, single_phylum, single_class, single_order, single_family, single_genus, single_spp)
```

## now going to use my custom salmon db to update those ASVs 

## output a list of all the salmon ASVs
```{r}
salmon <- sorted_tax_df %>%
  filter(genus == "Oncorhynchus") %>%
  select(qseqid) %>%
  unique()
```

## filter fasta to be just the salmon
```{r}
# Load the necessary libraries
library(Biostrings)

# Define the input and output file paths
fasta_file <- "/home/kimberly.ledger/nearshore_eDNA/data/samples_w_mock/filtered.seqs.fasta"
output_file <- "/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mock/oncorhynchus.fasta"

# Read the FASTA file
fasta_seqs <- readDNAStringSet(fasta_file)

# Create a list of ASV sequence headers
asv_headers <- salmon$qseqid  

# Filter the FASTA sequences based on the ASV headers
salmon_seqs <- fasta_seqs[names(fasta_seqs) %in% asv_headers]

# Write the filtered sequences to a new FASTA file
writeXStringSet(salmon_seqs, filepath = output_file)
```

okay, so i cleaned up reference seq and re-blast the salmon reads. not many changes were there so just manually going to enter in the IDs

```{r}
sorted_tax_df_2 <- sorted_tax_df %>%
  mutate(taxon = ifelse(qseqid == "ASV_0020", "Oncorhynchus keta", taxon),
         species = ifelse(qseqid == "ASV_0020", "Oncorhynchus keta", species),
         taxonomic_level = ifelse(qseqid == "ASV_0020", "species", taxonomic_level),
         taxon = ifelse(qseqid == "ASV_0042", "Oncorhynchus keta", taxon),
         species = ifelse(qseqid == "ASV_0042", "Oncorhynchus keta", species),
         taxonomic_level = ifelse(qseqid == "ASV_0042", "species", taxonomic_level),
         taxon = ifelse(qseqid == "ASV_0088", "Oncorhynchus mykiss", taxon),
         species = ifelse(qseqid == "ASV_0088", "Oncorhynchus mykiss", species),
         taxonomic_level = ifelse(qseqid == "ASV_0088", "species", taxonomic_level),
         taxon = ifelse(qseqid == "ASV_0458", "Oncorhynchus keta", taxon),
         species = ifelse(qseqid == "ASV_0458", "Oncorhynchus keta", species),
         taxonomic_level = ifelse(qseqid == "ASV_0458", "species", taxonomic_level)) %>%
  unique()
```

also going to made any pollock/pcod to just be Gadus 
```{r}
sorted_tax_df_3 <- sorted_tax_df_2 %>%
  mutate(taxon = ifelse(genus == "Gadus", "Gadus", taxon),
         taxonomic_level = ifelse(genus == "Gadus", "genus", taxonomic_level))
```


```{r}
write.csv(sorted_tax_df_3, "/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mock/tax_df_20240822.csv")
```
 

STOPPED HERE !!! 


# now let's take a closer look at the assignments we are getting from inital blastn 



okay, now that out-of-range species have been removed, i need to remake the taxonomic table to reflect the more specific ids

## so now determine whether ASVs should be assigned to species genus, family, order, etc. 
```{r}
#remove taxon_level and taxon designations
tax_df_2 <- tax_df_2 %>%
  select(!taxonomic_level) %>%
  select(!taxon) %>%
  filter(diff < 0.5)   ### go with 0.5% assignment threshold

singletons <- tax_df_2 %>%
  select(qseqid) %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1)

# these are the seqs that now have only a single match
singleton_df <- singletons %>%
  left_join(tax_df_2) %>%
  select(-n) %>%
  mutate(taxonomic_level = "species") %>%
  mutate(taxon = species)

## Genus-level matches
# remove the singletons from the bigger df 
single_genus <- tax_df_2 %>%
  anti_join(singleton_df) %>% # at best, these should be genus-level matches
  group_by(qseqid, genus) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single genus
  select(-n) %>%
  left_join(., tax_df_2) %>%
  mutate(taxonomic_level = "genus") %>%
  mutate(taxon = genus)
  
## Family-level matches
single_family <- tax_df_2 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  group_by(qseqid, family) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single family
  select(-n) %>%
  left_join(., tax_df_2) %>%
  mutate(taxonomic_level = "family") %>%
  mutate(taxon = family)


## Order-level matches
single_order <- tax_df_2 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  group_by(qseqid, order) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>%
  filter(n == 1) %>% # seqs that match a single order
  select(-n) %>%
  left_join(., tax_df_2) %>%
  mutate(taxonomic_level = "order") %>%
  mutate(taxon = order)

## Class-level matches
single_class <- tax_df_2 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  group_by(qseqid, class) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single class
  select(-n) %>%
  left_join(., tax_df_2) %>%
  mutate(taxonomic_level = "class") %>%
  mutate(taxon = class)

## Phylum-level matches
single_phylum <- tax_df_2 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, phylum) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single phylum
  select(-n) %>%
  left_join(., tax_df_2) %>%
  mutate(taxonomic_level = "phylum") %>%
  mutate(taxon = phylum)

# Kingdom level matches
single_kingdom <- tax_df_2 %>%
  anti_join(singleton_df)%>%
  anti_join(single_genus) %>%
  anti_join(single_family) %>%
  anti_join(single_order) %>%
  anti_join(single_class) %>%
  group_by(qseqid, kingdom) %>%
  tally() %>%
  ungroup() %>%
  group_by(qseqid) %>%
  tally() %>% 
  filter(n == 1) %>% # seqs that match a single kingdom
  select(-n) %>%
  left_join(., tax_df_2) %>%
  mutate(taxonomic_level = "kingdom") %>%
  mutate(taxon = kingdom)
```

## modify the singleton_df to include the right variable headers
```{r}
single_spp <- singleton_df
```

## recombine the full data set now that the appropriate level of taxonomy has been determined
```{r}
filtered_tax_df <- bind_rows(single_kingdom, single_phylum, single_class, single_order, single_family, single_genus, single_spp)
```


just so i can pick up on this without having to rerun all the code, let me export the filtered_tax_df
```{r}
#write.csv(filtered_tax_df, "/home/kimberly.ledger/nearshore_eDNA/outputs/samples_w_mock/tax_df_20240821.csv")
```
 
 
### STOPPED HERE DURING RERUN W/ MOCK COMMUNITIES INCLUDED ###  
 
## output a list of all the Gadidae ASVs
```{r}
gadidae <- filtered_tax_df %>%
  filter(family == "Gadidae") %>%
  select(qseqid) %>%
  unique()
```

## output a list of all the salmon ASVs
```{r}
salmon <- filtered_tax_df %>%
  filter(genus == "Oncorhynchus") %>%
  select(qseqid) %>%
  unique()
```

## filter fasta to be just the gadids 
```{r}
# Load the necessary libraries
library(Biostrings)

# Define the input and output file paths
fasta_file <- "/home/kimberly.ledger/nearshore_eDNA/data/filtered.seqs.fasta"
output_file <- "/home/kimberly.ledger/nearshore_eDNA/outputs/gadidae.fasta"

# Read the FASTA file
fasta_seqs <- readDNAStringSet(fasta_file)

# Create a list of ASV sequence headers
asv_headers <- gadidae$qseqid  # 

# Filter the FASTA sequences based on the ASV headers
gadidae_seqs <- fasta_seqs[names(fasta_seqs) %in% asv_headers]

# Write the filtered sequences to a new FASTA file
#writeXStringSet(gadidae_seqs, filepath = output_file)
```

## filter fasta to be just the salmon
```{r}
# Load the necessary libraries
library(Biostrings)

# Define the input and output file paths
fasta_file <- "/home/kimberly.ledger/nearshore_eDNA/data/filtered.seqs.fasta"
output_file <- "/home/kimberly.ledger/nearshore_eDNA/outputs/oncorhynchus.fasta"

# Read the FASTA file
fasta_seqs <- readDNAStringSet(fasta_file)

# Create a list of ASV sequence headers
asv_headers <- salmon$qseqid  

# Filter the FASTA sequences based on the ASV headers
salmon_seqs <- fasta_seqs[names(fasta_seqs) %in% asv_headers]

# Write the filtered sequences to a new FASTA file
writeXStringSet(salmon_seqs, filepath = output_file)
```


okay, so i cleaned up reference seq and re-blast the gadidae and salmon reads. let me compare the assignments 

```{r}
salmon_custom <- read.csv("/home/kimberly.ledger/nearshore_eDNA/outputs/oncorhynchus_asv_id.csv") %>%
  select(!X)

salmon_original <- filtered_tax_df %>%
  filter(genus == "Oncorhynchus") %>%
  select(qseqid, taxon, taxonomic_level) %>%
  unique()

salmon_join <- salmon_original %>%
  left_join(salmon_custom, by = "qseqid") %>%
  arrange(qseqid)
```

okay, so the cleaned up database doesn't get everything to species, but it surely does better. 

```{r}
gadidae_custom <- read.csv("/home/kimberly.ledger/nearshore_eDNA/outputs/gadidae_asv_id.csv") %>%
  select(!X) %>%
  mutate(taxon = ifelse(taxon == "Gadid", "Gadus", taxon))  ### whoops put gadid instead of gadus in table 
  
### i missed one ASV when doing the custom blast
gadidae_custom[14,1] <- "ASV_0183"
gadidae_custom[14,2] <- "Eleginus gracilis"
gadidae_custom[14,3] <- "species"

gadidae_original <- filtered_tax_df %>%
  filter(family == "Gadidae") %>%
  select(qseqid, taxon, taxonomic_level) %>%
  unique()

gadidae_join <- gadidae_original %>%
  left_join(gadidae_custom, by = "qseqid") %>%
  arrange(qseqid)
```


for this updated version of taxonomic ID i will go with the blast results using the custom db
```{r}
gad_sal_tax_df <- gadidae_custom %>%
  bind_rows(salmon_custom)
```

now i still need to consider the rest of the IDs 

these are the single species (excluding gadidae and oncorhynchus)
```{r}
single_species <- filtered_tax_df %>%
  filter(taxonomic_level == "species") %>%
  filter(family != "Gadidae") %>%
  filter(genus != "Oncorhynchus")
```

export for checking
```{r}
temp <- single_species %>%
  select(taxon, taxonomic_level) %>%
  unique()

#write.csv(single_species, "/home/kimberly.ledger/nearshore_eDNA/outputs/single_spp_taxon.csv", row.names = F)

#write.csv(temp, "/home/kimberly.ledger/nearshore_eDNA/outputs/single_spp.csv", row.names = F)
```


for a few groups, let's just consider the species hits that are < 0.5% of the best match 
```{r}
sebastidae <- filtered_tax_df %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Sebastidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(sebastidae, "/home/kimberly.ledger/nearshore_eDNA/outputs/sebastidae_pre.csv", row.names = F)
```

```{r}
pleuronectidae <- filtered_tax_df %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Pleuronectidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(pleuronectidae, "/home/kimberly.ledger/nearshore_eDNA/outputs/pleuronectidae_pre.csv", row.names = F)
```

```{r}
cottidae <- filtered_tax_df %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Cottidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(cottidae, "/home/kimberly.ledger/nearshore_eDNA/outputs/cottidae_pre.csv", row.names = F)
```

```{r}
stichaeidae <- filtered_tax_df %>%
  filter(taxonomic_level != "species")%>%
  filter(family == "Stichaeidae") %>%
  filter(diff < 0.5)

#export to make custom groups 
write.csv(stichaeidae, "/home/kimberly.ledger/nearshore_eDNA/outputs/stichaeidae_pre.csv", row.names = F)
```


okay, now just need to check out the missing tax ids 
```{r}
remaining <- filtered_tax_df %>%
  filter(taxonomic_level != "species") %>%
  filter(family != "Gadidae") %>%
  filter(genus != "Oncorhynchus") %>%
  filter(family != "Sebastidae") %>%
  filter(family != "Pleuronectidae") %>%
  filter(family != "Cottidae") %>%
  filter(family != "Stichaeidae")

#export for custom grouping 
write.csv(remaining, "/home/kimberly.ledger/nearshore_eDNA/outputs/remaining_pre.csv", row.names = F)
```

TRANSISTIONED HERE TO WORK IN EXCEL

read back in sebastidae, pleuro, and cottidae
```{r}
sebastidae_post <- read.csv("/home/kimberly.ledger/BeringSea_Arctic_eDNA/outputs/sebastidae_post.csv")
pleuronectidae_post <- read.csv("/home/kimberly.ledger/BeringSea_Arctic_eDNA/outputs/pleuronectidae_post.csv")
cottidae_post <- read.csv("/home/kimberly.ledger/BeringSea_Arctic_eDNA/outputs/cottidae_post.csv")
remaining_post <- read.csv("/home/kimberly.ledger/BeringSea_Arctic_eDNA/outputs/remaining_post.csv")
```


add the single species and the families back to the final tax df 
```{r}
final_tax_df <- single_species %>%
  bind_rows(sebastidae_post) %>%
  bind_rows(pleuronectidae_post) %>%
  bind_rows(cottidae_post) %>%
  bind_rows(remaining_post) %>%
  select(qseqid, taxon, taxonomic_level) %>%
  bind_rows(gad_sal_tax_df) %>%
  unique()
```

double check i didn't forget any ASVs in this process
```{r}
filtered_unique <- filtered_tax_df %>%
  select(qseqid) %>%
  unique() 

# Check if all elements in table1$qseqid are in table2$qseq
match <- all(filtered_unique$qseqid %in% final_tax_df$qseqid)

# Print result
if (match) {
  print("All rows in 'qseqid' column of table 1 match the rows in 'qseqid' column of table 2.")
} else {
  print("Not all rows in 'qseqid' column of table 1 match the rows in 'qseqid' column of table 2.")
}
```

## Create output taxonomy data frames
```{r}
# and then collapse that down to just a single taxon per ASV
final_tax_df <- final_tax_df %>%
  rename(ASV = qseqid)
```


```{r}
#write.csv(final_tax_df, "/home/kimberly.ledger/BeringSea_Arctic_eDNA/outputs/taxonomy_20240709.csv")
```

how many unique taxa? 
```{r}
length(unique(final_tax_df$taxon))
unique(final_tax_df$taxon)
```
